internal/worker/sshserver/worker.go

```
@@ -133,32 +133,14 @@ func (ssw *serverWrapperWorker) loop() error {
 		case <-ssw.catacomb.Dying():
 			return ssw.catacomb.ErrDying()
 		case <-changesChan:
-			// Restart the server worker.
-			srv.Kill()
-			if err := srv.Wait(); err != nil {
-				return errors.Trace(err)
-			}
-
-			// Get latest controller configuration.
-			port, _, err := ssw.getLatestControllerConfig()
-			if err != nil {
-				return errors.Trace(err)
-			}
-
-			// Start the server again.
-			srv, err = ssw.config.NewServerWorker(ServerWorkerConfig{
-				Logger:      ssw.config.Logger,
-				JumpHostKey: jumpHostKey,
-				Port:        port,
-			})
+			newPort, _, err := ssw.getLatestControllerConfig()
 			if err != nil {
-				return errors.Trace(err)
+				continue
```

what if it's a fatal error? it continues restarting and we wouldn't know what's happening. If we continue on this error I would add a log.

---

Yeah true, we can return the error here actually.

---

Yes; return the error here. It will cascade up and kill the parent worker.

You can rely on the fact that if there is a non-nil error, unrecognised by the dependency engine, it will restart your worker tree from the parent anyway.